### PR TITLE
III-6435 Add projector to track event place changes

### PR DESCRIPTION
**Pull request by grubolsch**
### Added
Add projector to track location changes for Places.

I track when events get created, when events get moved (UpdateLocation), and when copies are copied.

---

Ticket: https://jira.uitdatabank.be/browse/III-6435

### CODE REVIEW

File: app/Event/EventReadServiceProvider.php [hunk 1]
```diff
@@ -52,5 +59,23 @@ function () use ($container): JSONLDMainLanguageQuery {
                 );
             }
         );
+
+        $container->addShared(
+            EventLocationHistoryProjector::class,
```
Comment (LucWollants) [inline app/Event/EventReadServiceProvider.php:64]: If you look at other projectors they all use `Place` in their name, for example `PlaceRelationsProjector`. If you think about it, this is the `PlaceRelationsHistoryProjector`

File: app/Migrations/Version20241129150800.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+final class Version20241129150800 extends AbstractMigration
+{
+    public function up(Schema $schema): void
+    {
+        $table = $schema->createTable('event_location_history');
+
+        $table->addColumn('id', 'integer', ['autoincrement' => true]);
```
Comment (LucWollants) [inline app/Migrations/Version20241129150800.php:16]: Do we need this id?

File: src/Model/ValueObject/Identity/UUID.php [hunk 3]
```diff
@@ -27,4 +27,9 @@ public function __construct(string $value)
 
         $this->setValue($value);
     }
+
+    public static function uuid4(): self
```
Comment (LucWollants) [inline src/Model/ValueObject/Identity/UUID.php:31]: I think it's tricky to place that here, look at the comment inside the constructor. It also supports another legacy format.

Also if you would like to add this, please add it in another pull request and modify all existing `Uuid::uuid4()` to `UUID::uuid4()` because you did not take into account that PHP ignores casing.

File: app/Event/EventReadServiceProvider.php (hunk 1 above)
Comment (grubolsch) [reply app/Event/EventReadServiceProvider.php:64]: I would keep `event` in the name because is the relation between `Event` and `Place` that is being tracked.
It also listens to Event related commands like `EventCreated`, not place related events.
Can we land on "EventRelationsPlaceHistoryProjector"?

I put Event first because it is the relation from the viewpoint of an event we track.

File: src/Event/ReadModel/History/EventLocationHistoryProjector.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        $myEvent = $this->eventRepository->fetch($eventId);
+
+        $body = $myEvent->getAssocBody();
+
+        $url = $body['location']['@id'];
+
+        // Return everything after the last slash
+        // https://io.uitdatabank.be/event/fd7dbfaf-5162-4181-9446-e61c4f5ef3f2 -> fd7dbfaf-5162-4181-9446-e61c4f5ef3f2
+        return new UUID(substr($url, strrpos($url, '/') + 1));
```
Comment (LucWollants) [inline src/Event/ReadModel/History/EventLocationHistoryProjector.php:82]: I think we have this class for it `RegexUUIDParser`

File: tests/Event/ReadModel/Relations/Doctrine/DBALEventLocationHistoryRepositoryTest.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    /** @test  */
+    public function should_store_event_location_starting_point(): void
+    {
+        $eventId = Uuid::uuid4();
+        $placeId = Uuid::uuid4();
+
+        $this->repository->storeEventLocationStartingPoint($eventId, $placeId);
+
+        $result = $this->connection->fetchAssociative(
```
Comment (LucWollants) [inline tests/Event/ReadModel/Relations/Doctrine/DBALEventLocationHistoryRepositoryTest.php:31]: There is even no need for a `WHERE` clause, there is only one row.

File: app/Event/EventReadServiceProvider.php (hunk 1 above)
Comment (LucWollants) [reply app/Event/EventReadServiceProvider.php:64]: `EventRelationsPlaceHistoryProjector` is quiet long, maybe `EventPlaceHistoryProjector`?

File: app/Migrations/Version20241129150800.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        $table->addColumn(
+            'date',
+            'datetime_immutable',
+            ['notnull' => true]
+        );
+
+        $table->setPrimaryKey(['id']);
+
+        // Index because I assume one of the most asked questions is going to be: "where are all the events of place X?"
+        $table->addIndex(['old_place'], 'idx_old_place');
```
[the remark this answers is not in the document]
Comment (grubolsch) [reply app/Migrations/Version20241129150800.php]: Added a bunch of indexes, but we will have to see in practice what type of queries we will run to this table.

**Comment by grubolsch** [comment]
All feedback has been processed

File: tests/Event/samples/event_with_dummy_location.cdbxml.xml [hunk 7]
```diff
@@ -29,7 +29,7 @@
             <cdb:title>'Bekend met Gent' - quiz</cdb:title>
         </cdb:eventdetail>
     </cdb:eventdetails>
-    <cdb:location>
+    <cdb:location cdbid="">
```
Comment (JonasVHG) [inline tests/Event/samples/event_with_dummy_location.cdbxml.xml:32]: Why did you add `cdbid=""`? Such cdbXml's never existed. It either had a valid `cdbid` or now `cdbid` property.
This file should not be changed.

File: app/Event/EventReadServiceProvider.php [hunk 8]
```diff
@@ -52,5 +59,23 @@ function () use ($container): JSONLDMainLanguageQuery {
...
             }
         );
+
+        $container->addShared(
+            EventPlaceHistoryProjector::class,
+            function () use ($container): EventPlaceHistoryProjector {
+                return new EventPlaceHistoryProjector(
+                    $container->get(EventLocationHistoryRepository::class),
+                    $container->get('event_jsonld_cache'),
+                    LoggerFactory::create($this->getContainer(), LoggerName::forService('event_location_history'))
```
Comment (LucWollants) [inline app/Event/EventReadServiceProvider.php:69]: Not sure if we need a dedicated logger for this, most of the simple HTTP requests and their command and event handling or just logged inside `web.log`

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+class EventPlaceHistoryProjector implements EventListener
+{
+    use DelegateEventHandlingToSpecificMethodTrait;
+
+    private EventLocationHistoryRepository $repository;
+    private DocumentRepository $eventRepository;
+    private LoggerInterface $logger;
+
+    public function __construct(
+        EventLocationHistoryRepository $eventLocationHistoryRepository,
```
Comment (LucWollants) [inline src/Event/ReadModel/History/EventPlaceHistoryProjector.php:34]: The naming is still inconsistent, there is a mix of `EventPlaceHistoryProjector` and `EventLocationHistoryRepository`

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                $event->getCdbXml()
+            );
+        } catch (CultureFeed_Cdb_ParseException $e) {
+            $this->logger->error(sprintf('Failed to store event imported from UDB2: %s', $e->getMessage()));
+            return;
+        }
+
+        $this->repository->storeEventLocationStartingPoint(
+            new UUID($event->getEventId()),
+            new UUID($udb2Event->getLocation()->getCdbid())
```
Comment (LucWollants) [inline src/Event/ReadModel/History/EventPlaceHistoryProjector.php:96]: I think we need to take into account that a lot of older places have an XML without a Cdbid, this should be caught and not make the handler crash.

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        EventLocationHistoryRepository $eventLocationHistoryRepository,
+        DocumentRepository $eventRepository,
+        LoggerInterface $logger
+    ) {
+        $this->repository = $eventLocationHistoryRepository;
+        $this->eventRepository = $eventRepository;
+        $this->logger = $logger;
+    }
+
+    protected function applyLocationUpdated(LocationUpdated $event): void
```
Comment (LucWollants) [inline src/Event/ReadModel/History/EventPlaceHistoryProjector.php:43]: It's sometimes hard to figure out in this context if `$event` is the resource or the item from the event store. Would it be easier to understand if `$locationUpdated` is used?

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            $newPlaceId
+        );
+    }
+
+    protected function applyMajorInfoUpdated(MajorInfoUpdated $event): void
+    {
+        try {
+            $oldPlaceId = $this->getOldPlaceUuid($event->getItemId());
+        } catch (DocumentDoesNotExist $e) {
+            $this->logger->error(sprintf('Failed to store location updated: %s', $e->getMessage()));
```
Comment (LucWollants) [inline src/Event/ReadModel/History/EventPlaceHistoryProjector.php:137]: Copy/paste error, this is `MajorInfoUpdated`

File: src/Event/ReadModel/Relations/Doctrine/DBALEventLocationHistoryRepository.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        $this->insertIntoLocationHistoryTable($eventId, null, $placeId);
+    }
+
+    public function storeEventLocationMove(UUID $eventId, UUID $oldPlaceId, UUID $newPlaceId): void
+    {
+        $this->insertIntoLocationHistoryTable($eventId, $oldPlaceId, $newPlaceId);
+    }
+
+    public function insertIntoLocationHistoryTable(UUID $eventId, ?UUID $oldPlaceId, UUID $newPlaceId): void
```
Comment (LucWollants) [inline src/Event/ReadModel/Relations/Doctrine/DBALEventLocationHistoryRepository.php:31]: Weird method name if the table name is `event_location_history`

File: tests/Event/ReadModel/History/EventPlaceHistoryProjectorTest.php [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+            ->method('storeEventLocationMove')
+            ->with(
+                $eventId,
+                $oldPlaceId,
+                $newPlaceId
+            );
+
+        $eventUpdatedFromUDB2 = new EventUpdatedFromUDB2(
+            $eventId->toString(),
+            SampleFiles::read(__DIR__ . '/../../samples/event_with_existing_location.cdbxml.xml'),
```
Comment (LucWollants) [inline tests/Event/ReadModel/History/EventPlaceHistoryProjectorTest.php:265]: I think you should also write a test where the location is a dummy location. There is a sample file for this `tests/Event/samples/event_with_dummy_location.cdbxml.xml`

File: app/Migrations/Version20241129150800.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+            'notnull' => false,
+        ]);
+        $table->addColumn('new_place', 'guid', ['length' => 36]);
+        $table->addColumn(
+            'date',
+            'datetime_immutable',
+            ['notnull' => true]
+        );
+
+        $table->addIndex(['old_place', 'date', 'event'], 'idx_old_place');
```
[the remark this answers is not in the document]
Comment (grubolsch) [reply app/Migrations/Version20241129150800.php:27]: I kinda agree, I am just guessing here.
Do I just drop all indexes? Or keep a specific one?

File: tests/Event/samples/event_with_dummy_location.cdbxml.xml (hunk 7 above)
Comment (grubolsch) [reply tests/Event/samples/event_with_dummy_location.cdbxml.xml:32]: A commit mistake, I reverted it!

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php (hunk 9 above)
Comment (grubolsch) [reply src/Event/ReadModel/History/EventPlaceHistoryProjector.php]: I made it consistent, and then also made methods / the table name consistent.

File: src/Event/ReadModel/History/EventPlaceHistoryProjector.php (hunk 11 above)
Comment (grubolsch) [reply src/Event/ReadModel/History/EventPlaceHistoryProjector.php]: Renamed all the variables

File: app/Migrations/Version20241129150800.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (LucWollants) [reply app/Migrations/Version20241129150800.php:27]: Feel free to drop them all

File: tests/Event/ReadModel/History/EventPlaceHistoryProjectorTest.php (hunk 14 above)
Comment (grubolsch) [reply tests/Event/ReadModel/History/EventPlaceHistoryProjectorTest.php:265]: Added this, with using `cdbIdExtractor::getRelatedPlaceCdbId()`

**Comment by grubolsch** [comment]
All feedback has been processed

**Review by LucWollants — APPROVED** [review]
Nice work! Thanks for all the extra refactor work and using consistent naming
